ThirdpartyToolchain.cmake

```diff
@@ -0,0 +1,127 @@
+# include(ThirdpartyToolchain) resolve_dependency(folly)
+#
+# ========================================================================================
+
+include(FetchContent)
+
+# =====================================FOLLY==============================================
+
+if(DEFINED ENV{VELOX_FOLLY_URL})
+  set(FOLLY_SOURCE_URL "$ENV{VELOX_FOLLY_URL}")
+else()
+  set(VELOX_FOLLY_BUILD_VERSION v2022.07.11.00)
+  set(FOLLY_SOURCE_URL
+      "https://github.com/facebook/folly/archive/${VELOX_FOLLY_BUILD_VERSION}.tar.gz"
+  )
+  set(VELOX_FOLLY_BUILD_SHA256_CHECKSUM
+      b6cc4082afd1530fdb8d759bc3878c1ea8588f6d5bc9eddf8e1e8abe63f41735)
+endif()
+
+macro(build_folly)
+  message(STATUS "Building Folly from source")
+  # FOLLY_CXX_FLAGS is used internally on folly to define CMAKE_CXX_FLAGS.
+  # Define some known warnings to avoid possible errors on some OS/archs
+  set(FOLLY_CXX_FLAGS -Wno-nullability-completeness -Wno-unused
+                      -Wno-unused-parameter -Wno-overloaded-virtual)
+  FetchContent_Declare(
+    folly
+    URL ${FOLLY_SOURCE_URL}
+    URL_HASH SHA256=${VELOX_FOLLY_BUILD_SHA256_CHECKSUM})
+  if(NOT folly_POPULATED)
+    # Fetch the content using previously declared details
+    FetchContent_Populate(folly)
+    add_subdirectory(${folly_SOURCE_DIR} ${folly_BINARY_DIR})
+    # Avoid possible errors for known warning if SSL>3 is used
+    target_compile_options(folly PUBLIC -Wno-deprecated-declarations)
+  endif()
+  set(FOLLY_BENCHMARK_STATIC_LIB
+      ${folly_BINARY_DIR}/folly/libfollybenchmark${CMAKE_STATIC_LIBRARY_SUFFIX})
+  set(FOLLY_LIBRARIES folly)
+endmacro()
+# ===============================END FOLLY================================
+
+macro(build_dependency DEPENDENCY_NAME)
+  if("${DEPENDENCY_NAME}" STREQUAL "folly")
+    build_folly()
+  else()
+    message(
+      FATAL_ERROR "Unknown thirdparty dependency to build: ${DEPENDENCY_NAME}")
+  endif()
+endmacro()
+
+# * Macro to resolve thirparty dependencies.
+#
+# Provides the macro resolve_dependency(). This macro will allow us to find the
+# dependency via the usage of find_package or use the custom
+# build_dependency(DEPENDENCY_NAME) macro to download and build the third party
+# dependency.
+#
+# resolve_dependency(DEPENDENCY_NAME [REQUIRED_VERSION <required_version>] )
+#
+# The resolve_dependency() macro can be used to define a thirdparty dependency.
+#
+# ${DEPENDENCY_NAME}_SOURCE is expected to be set to either AUTO, SYSTEM or
+# BUNDLED. If ${DEPENDENCY_NAME}_SOURCE is SYSTEM it will try to find the
+# corresponding package via find_package and if not found it will call the
+# build_dependency macro to download and build the third party dependency. If
+# ${DEPENDENCY_NAME}_SOURCE is SYSTEM it will force to find via find_package. If
+# ${DEPENDENCY_NAME}_SOURCE is BUNDLED it will force to build from source.
+#
+# If REQUIRED_VERSION is provided it will be used as the VERSION to be used on
```

Comment:
If REQUIRED_VERSION is set and SYSTEM is source, will it fail or download if SYSTEM doesnt meet version requirements.

Reply:
Thanks yes, I agree with this behavior.